### PR TITLE
Geometric Deep Learning: VTK dataloader, mesh utilities, file format converters

### PR DESCRIPTION
**Pull request by mnabian**
# Modulus Pull Request

This PR introduces the MeshDatapipe class, which integrates NVIDIA DALI to efficiently process vtp, vtu, cgns files within PyTorch pipelines. Key features include support for lazy-loading and normalization of attributes using pre-computed statistics, configurable options for batch size, device selection, and parallel processing for scalable multi-process environments. The update aims to enhance data processing performance by leveraging DALI's GPU capabilities, providing a streamlined solution for handling large mesh datasets in machine learning and simulation applications. Additionally, the PR includes robust error handling, new helper functions for VTK data parsing, and thorough documentation to guide users on setting up and using the pipeline.

## Description

Closes https://github.com/NVIDIA/modulus/issues/500

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+        # when corresponding pipeline/dataset is destroyed.
+        data = self.vtk_reader_fn(self.data_paths[idx])
+        # vertices, pressure, wss = self.parse_vtkpolydata(polydata)
+
+        return self.parse_vtk_data_fn(data, self.vars)
+
+    def __len__(self):
+        return len(self.indices)
+
+    def vtk_reader(self):
```
Comment (ktangsali) [inline modulus/datapipes/geometric/vtk_datapipe.py:357]: Now that we have cgns too, can we name this something generic?

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+            raise ValueError(
+                f"Failed to get array {array_name} from the unstructured grid."
+            )
+        array_data = np.zeros(
+            (points.GetNumberOfPoints(), array.GetNumberOfComponents())
+        )
+        for j in range(points.GetNumberOfPoints()):
+            array.GetTuple(j, array_data[j])
+        attributes.append(torch.tensor(array_data, dtype=torch.float32))
+        print(array_data[1002, :])
```
Comment (ktangsali) [inline modulus/datapipes/geometric/vtk_datapipe.py:456]: Clean-up needed

File: modulus/utils/geometric/convert_file_formats.py
Comment (ktangsali) [inline modulus/utils/geometric/convert_file_formats.py:1]: Would be good to add tests for all these utils

File: modulus/datapipes/geometric/vtk_datapipe.py
Comment (ktangsali) [inline modulus/datapipes/geometric/vtk_datapipe.py:1]: Good to add tests for this datapipe

File: modulus/utils/geometric/convert_file_formats.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    Args:
+    - input_file (str): Path to the input VTP file.
+    - output_file (str): Path to save the converted STL file.
+    """
+    reader = vtk.vtkXMLPolyDataReader()
+    reader.SetFileName(input_file)
+    if not reader.CanReadFile(input_file):
+        raise ValueError(f"Error: Could not read file: {input_file}")
+    reader.Update()
```
Comment (ktangsali) [inline modulus/utils/geometric/convert_file_formats.py:57]: Don't you have to extract surfaces alone? Or alteast check that the loaded vtp is a surface data? 

I thought something like this was required:

```
    # Convert unstructured grid to polydata.
    surface_filter = vtk.vtkDataSetSurfaceFilter()
    surface_filter.SetInputConnection(reader.GetOutputPort())
    surface_filter.Update()
```

File: modulus/utils/geometric/convert_file_formats.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        converter = convert_obj_to_vtp
+    elif conversion_type == "vtp2stl":
+        src_ext = ".vtp"
+        dst_ext = ".stl"
+        converter = convert_vtp_to_stl
+    else:
+        raise NotImplementedError(
+            f"Conversion type {conversion_type} is not supported."
+        )
+
```
Comment (ktangsali) [inline modulus/utils/geometric/convert_file_formats.py:88]: Same comment as above regarding volumetric data. Please check if this will work with volumetric data (I think it might not, in which case the scope needs to be clarified.)

**Review by ktangsali — COMMENTED** [review]
The utilities look good overall, but this PR is missing a few things:
1. A description of the PR would be helpful to clarify the intent.
2. Tests for the various utils are required.

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+        attributes.append(torch.tensor(array_data, dtype=torch.float32))
+    attributes = torch.cat(attributes, dim=-1)
+
+    # Fetch edges
+    polys = polydata.GetPolys()
+    if polys is None:
+        raise ValueError("Failed to get polygons from the polydata.")
+    polys.InitTraversal()
+    edges = []
+    for i in range(polys.GetNumberOfCells()):
```
Comment (ktangsali) [inline modulus/datapipes/geometric/vtk_datapipe.py:416]: Same comment as here: https://github.com/NVIDIA/modulus/pull/544#discussion_r1635177257

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+            array = point_data.GetArray(array_name)
+        except ValueError:
+            raise ValueError(f"Failed to get array {array_name} from the unstructured grid.")
+        array_data = np.zeros(
+            (points.GetNumberOfPoints(), array.GetNumberOfComponents())
+        )
+        for j in range(points.GetNumberOfPoints()):
+            array.GetTuple(j, array_data[j])
+        attributes.append(torch.tensor(array_data, dtype=torch.float32))
+        print(array_data[1002,:])
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py]: Done

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+        std_stat_file = self.stats_dir / Path("global_stds.npy")
+
+        if not mean_stat_file.exists():
+            raise IOError(f"Mean statistics file {mean_stat_file} not found")
+        if not std_stat_file.exists():
+            raise IOError(f"Std statistics file {std_stat_file} not found")
+
+        # has shape [1, C]
+        self.mu = np.load(str(mean_stat_file))[:, 0 : self.num_vars]
+        # has shape [1, C]
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py:204]: Fixed it in line 179

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 2 above)
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py:456]: Done

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 5 above)
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py:416]: I'm not sure about this. Maybe we should keep an eye on this. I checked the number of cells and it exactly matches with the input data.

File: modulus/utils/geometric/convert_file_formats.py (hunk 3 above)
Comment (mnabian) [reply modulus/utils/geometric/convert_file_formats.py:57]: Clarified the scope

File: modulus/utils/geometric/convert_file_formats.py (hunk 4 above)
Comment (mnabian) [reply modulus/utils/geometric/convert_file_formats.py:88]: Clarified the scope

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+class VTKDatapipe(Datapipe):
+    """DALI data pipeline for VTK files
+
+    Parameters
+    ----------
+    data_dir : str
+        Directory where ERA5 data is stored
+    vars : List[str, None]
+        Ordered list of variables to be loaded from the files
+    num_vars : int
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py:63]: It becomes a bit messy to infer it from the `vars` argument. FOr that the only way is to get it from the external source as a numpy array, but it'll compute it every time the external source is called. I think specifying that explicitly is a better approach.

File: modulus/datapipes/geometric/vtk_datapipe.py [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+class MetaData(DatapipeMetaData):
+    name: str = "VTK"
+    # Optimization
+    auto_device: bool = True
+    cuda_graphs: bool = True
+    # Parallel
+    ddp_sharding: bool = True
+
+
+class VTKDatapipe(Datapipe):
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py:54]: Added, thanks

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 5 above)
Comment (ktangsali) [reply modulus/datapipes/geometric/vtk_datapipe.py]: I don't think the cells is the issue here, but rather the edges. Again, might be ok for training a GNN, but if we don't capture all the edges, that sounds a bit problematic to me.

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ktangsali) [reply modulus/datapipes/geometric/vtk_datapipe.py]: I feel like there two are repetitive and can create confusion. The only time these two will not be the same is when we have vector quantities in the `vars`. Is that correct? Or do we expect any other use case where the two will not be same?

File: modulus/datapipes/geometric/vtk_datapipe.py
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py:1]: Done

File: modulus/utils/geometric/convert_file_formats.py
Comment (mnabian) [reply modulus/utils/geometric/convert_file_formats.py:1]: Added unit tests

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py]: That is correct, these won't be the same if there are vector quantities. For this reason and also for implementation purposes, I preferred to have the number of variables be explicitly defined.

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 5 above)
Comment (mnabian) [reply modulus/datapipes/geometric/vtk_datapipe.py]: I fixed this according to your suggested fix for the stokes flow example.

File: test/datapipes/test_mesh_datapipe.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    vtu_writer = vtk.vtkXMLUnstructuredGridWriter()
+    vtu_writer.SetFileName(os.path.join(dir, "random.vtu"))
+    vtu_writer.SetInputData(vtu_mesh)
+    vtu_writer.Write()
+
+
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+def test_mesh_datapipe(device):
+    """Tests the MeshDatapipe class with VTP and VTU files."""
+    dir = "temp_data"
```
Comment (Alexey-Kamenev) [inline test/datapipes/test_mesh_datapipe.py:107]: Might want to use `tmp_path` fixture instead.

Also, `dir` is a Python built-in.

File: test/datapipes/test_mesh_datapipe.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+    # Remove the directory
+    shutil.rmtree(dir)
+
+
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+def test_mesh_datapipe_cgns(device):
+    """Tests the mesh datapipe for CGNS file format."""
+    datapipe_cgns = MeshDatapipe(
+        data_dir="/code/PRs/geometric/modulus-data/datasets/sample_formats",
```
Comment (Alexey-Kamenev) [inline test/datapipes/test_mesh_datapipe.py:153]: Hmm, will this path pass our CI?

**Review by Alexey-Kamenev — APPROVED** [review]
Mostly minor comments.

File: test/datapipes/test_mesh_datapipe.py (hunk 11 above)
Comment (mnabian) [reply test/datapipes/test_mesh_datapipe.py:153]: Oh I forgot to change this after local tests, thanks!

File: modulus/datapipes/cae/mesh_datapipe.py [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterable, List, Tuple, Union
+
+from modulus.datapipes.datapipe import Datapipe
+from modulus.datapipes.meta import DatapipeMetaData
+
+from .readers import read_cgns, read_vtp, read_vtu
+
+Tensor = torch.Tensor
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/cae/mesh_datapipe.py:41]: Done

File: modulus/datapipes/cae/mesh_datapipe.py [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+    process_rank : int, optional
+        Rank ID of local process, by default 0
+    world_size : int, optional
+        Number of training processes, by default 1
+    """
+
+    def __init__(
+        self,
+        data_dir: str,
+        vars: List[str],
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/cae/mesh_datapipe.py:90]: Thanks, changed to `variables`.

File: modulus/datapipes/cae/mesh_datapipe.py [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+    def parse_dataset_files(self) -> None:
+        """Parses the data directory for valid files and determines training samples
+
+        Raises
+        ------
+        ValueError
+            In channels specified or number of samples per year is not valid
+        """
+        # get all input data files
+        if self.file_format == "vtp":
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/cae/mesh_datapipe.py:144]: This is pretty neat! Switched to Match Case statement

File: modulus/datapipes/cae/mesh_datapipe.py [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+                [str(path) for path in self.data_dir.glob("*.cgns")]
+            )
+        else:
+            raise NotImplementedError(
+                f"Data type {self.file_format} is not supported yet"
+            )
+
+        for data_path in self.data_paths:
+            self.logger.info(f"File found: {data_path}")
+        self.all_samples = len(self.data_paths)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/cae/mesh_datapipe.py:163]: I agree, changed to `total_samples`

File: modulus/datapipes/cae/mesh_datapipe.py [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+        # so there is no need to explicitly close the files - this will be done
+        # when corresponding pipeline/dataset is destroyed.
+        if self.data_cache[self.data_paths[idx]] is None:
+            data = self.mesh_reader_fn(self.data_paths[idx])
+            processed_data = self.parse_vtk_data_fn(data, self.vars)
+            self.data_cache[self.data_paths[idx]] = processed_data
+        else:
+            processed_data = self.data_cache[self.data_paths[idx]]
+
+        return self.parse_vtk_data_fn(data, self.vars)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/cae/mesh_datapipe.py:361]: You are absolutely right, and I adopted your suggestion

File: modulus/datapipes/cae/mesh_datapipe.py [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+            raise ValueError(
+                f"Failed to get array {array_name} from the unstructured grid."
+            )
+        array_data = np.zeros(
+            (points.GetNumberOfPoints(), array.GetNumberOfComponents())
+        )
+        for j in range(points.GetNumberOfPoints()):
+            array.GetTuple(j, array_data[j])
+        attributes.append(torch.tensor(array_data, dtype=torch.float32))
+    attributes = torch.cat(attributes, dim=-1)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply modulus/datapipes/cae/mesh_datapipe.py:417]: Added

File: test/datapipes/test_mesh_datapipe.py (hunk 10 above)
Comment (mnabian) [reply test/datapipes/test_mesh_datapipe.py:107]: Switched to using `tmp_path`, thanks!

File: test/datapipes/test_mesh_datapipe.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+        assert data[0]["vertices"].shape == (1, 20, 3)
+        assert data[0]["x"].shape == (1, 20, 1)
+
+
+@nfsdata_or_fail
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+def test_mesh_datapipe_cgns(device, pytestconfig):
+    """Tests the mesh datapipe for CGNS file format."""
+    datapipe_cgns = MeshDatapipe(
+        data_dir=cgns_data_dir,
```
Comment (Alexey-Kamenev) [inline test/datapipes/test_mesh_datapipe.py:158]: I think you forgot to add `cgns_data_dir` to the test function arguments? That's why the CI failed since `cgns_data_dir` is a fixture (function)?

File: test/datapipes/test_mesh_datapipe.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+    # Write VTU mesh to file
+    vtu_writer = vtk.vtkXMLUnstructuredGridWriter()
+    vtu_writer.SetFileName(os.path.join(dir, "random.vtu"))
+    vtu_writer.SetInputData(vtu_mesh)
+    vtu_writer.Write()
+
+
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+def test_mesh_datapipe(device, tmp_path):
```
Comment (ktangsali) [inline test/datapipes/test_mesh_datapipe.py:111]: Would be great if we can add the package check decorator for these similar to here: https://github.com/NVIDIA/modulus/blob/main/test/datapipes/test_ahmed_body.py#L33 to ensure smoother experience for the pip installers

File: modulus/datapipes/geometric/vtk_datapipe.py (hunk 5 above)
Comment (ktangsali) [reply modulus/datapipes/geometric/vtk_datapipe.py]: Great, thank you!

**Review by ktangsali — APPROVED** [review]
Looks good to me, thanks for addressing my comments. Cheers!

File: test/datapipes/test_mesh_datapipe.py (hunk 19 above)
Comment (mnabian) [reply test/datapipes/test_mesh_datapipe.py]: Done
